File: DailyTemperatures.cpp
// Time Complexity: O(n)
// Space Complexity: O(n)

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        deque<pair<int,int>> dq; 
        vector<int> result(temperatures.size(), 0); 

        dq.push_back({temperatures[0], 0});

        for(int i = 1; i < temperatures.size(); i++) {
            int currentTemperature = temperatures[i];
            while(!dq.empty() && currentTemperature > dq.back().first) {
                int temp = dq.back().first; 
                int index = dq.back().second;
                dq.pop_back();

                result[index] = i - index;  
            }

            dq.push_back({currentTemperature, i});
        }

        return result; 
                 
    }
};

File: src/DailyTemperatures.java
// Time Complexity:  O(n)
// Space Complexity: O(n)

class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] res = new int[n];
        Stack<Integer> stack = new Stack<>();
        for(int i=0; i<n; i++) {
            int temp = temperatures[i];                                        // take temperature one by one
            while(!stack.isEmpty() && temperatures[stack.peek()] < temp) {     // remove all temperature less than this one
                int ind = stack.pop();                           
                res[ind] = i-ind;                                              // put difference of indexes in result
            }
            stack.push(i);
        }
        return res;
    }
}
